File: DMMPVP/core/src/main/java/com/zenyte/game/content/drops/DropTableBuilder.java
package com.zenyte.game.content.drops;

import com.zenyte.game.content.drops.table.DropTable;

/**
 * @author Kris | 22/01/2020
 * @see <a href="https://www.rune-server.ee/members/kris/">Rune-Server profile</a>
 */
public class DropTableBuilder {
    
    private final DropTable table = new DropTable();
    
    public DropTableBuilder append(final int id) {
        return append(id, 1, 1, 1);
    }
    
    public DropTableBuilder append(final int id, final int weight) {
        return append(id, 1, 1, weight);
    }
    
    public DropTableBuilder append(final int id, final int amount, final int weight) {
        return append(id, amount, amount, weight);
    }
    
    public DropTableBuilder append(final int id, final int minAmount, final int maxAmount, final int weight) {
        table.append(id, weight, minAmount, maxAmount);
        return this;
    }
    
    public DropTable build() {
        return table;
    }
    
}

File: DMMPVP/scripts/npc/drops/src/main/kotlin/com/near_reality/scripts/npc/drops/DropTableBuilder.kt
package com.near_reality.scripts.npc.drops

import com.zenyte.game.world.entity.npc.drop.matrix.Drop
import com.zenyte.game.world.entity.npc.drop.matrix.NPCDrops
import it.unimi.dsi.fastutil.objects.ObjectArrayList
import it.unimi.dsi.fastutil.objects.ObjectList

/**
 * @author Jire
 */
@Deprecated("Use com.near_reality.scripts.npc.drops.table.dsl.DropTableBuilder")
class DropTableBuilder(
    private val npcID: Int,
    private val weight: Int = 0
) {

    private val drops: ObjectList<Drop> = ObjectArrayList()

    operator fun Int.invoke(
        rate: Int = Drop.GUARANTEED_RATE,
        minAmount: Int = MIN_AMOUNT_REQUIRED,
        maxAmount: Int = minAmount
    ) {
        check(minAmount >= MIN_AMOUNT_REQUIRED) { "Min amount ($minAmount) must be at least $MIN_AMOUNT_REQUIRED" }
        check(maxAmount >= MIN_AMOUNT_REQUIRED) { "Max amount ($maxAmount) must be at least $MIN_AMOUNT_REQUIRED" }
        check(minAmount <= maxAmount) { "Min amount ($minAmount) cannot be more than max amount ($maxAmount)" }

        drops.add(Drop(this, rate, minAmount, maxAmount))
    }
    operator fun Int.invoke(
        rate: Int = Drop.GUARANTEED_RATE,
        amountRange: IntRange
    )  = invoke(rate, amountRange.first, amountRange.last)

    fun buildTable() = NPCDrops.DropTable(npcID, weight, drops.sortedBy { it.baseRate }.toTypedArray())

    companion object {
        const val MIN_AMOUNT_REQUIRED = 1
    }

}
